resolwe/storage/manager.py

```diff
@@ -18,251 +18,307 @@
+                        connector_name=connector_name,
+                    )
+                )
+            )
+            .filter(any_location_exists=True, to_location_exists=False)
+            .distinct()
         )
 
-        if "delay" in copy_rules:
-            # Do not copy if delay is negative.
-            rule_results["delay_non_negative"] = copy_rules["delay"] >= 0
-
-            move_delay = timedelta(days=copy_rules["delay"])
-            current_delay = now() - self.file_storage.created
-            rule_results["delay"] = current_delay >= move_delay
-        return all(rule_results.values())
+    def delete_locations(self, connector_name: str) -> models.QuerySet:
+        """Get a list of locations that can safely be deleted."""
+        location_settings = copy.deepcopy(STORAGE_CONNECTORS.get(connector_name, {}))
+        rules: dict = location_settings.get("config", {}).get("delete", {})
 
-    def delete(self) -> Optional[StorageLocation]:
-        """Get StorageLocation object that can be deleted.
+        if not rules:
+            return FileStorage.objects.none()
 
-        When multiple StorageLocation objects can be deleted the one with
-        lowest priority is returned. When no StorageLocation objects can be
-        deleted None is returned.
+        process_overrides = rules.pop("process_type", {})
+        data_slug_overrides = rules.pop("data_slug", {})
 
-        :return: storage location that can be safely removed or None.
-        :rtype: Optional[StorageLocation]
-        """
-        can_delete = [
-            storage_location
-            for storage_location in self.file_storage.storage_locations.all()
-            if self._should_delete(storage_location)
+        whens = [
+            models.When(
+                data__slug=data_slug,
+                then=timedelta(days=self._preprocess_delay(override_rule.get("delay"))),
+            )
+            for data_slug, override_rule in data_slug_overrides.items()
+        ]
+        whens += [
+            models.When(
+                data__process__type__startswith=process_type,
+                then=timedelta(days=self._preprocess_delay(override_rule.get("delay"))),
+            )
+            for process_type, override_rule in process_overrides.items()
         ]
-        if can_delete:
-            return sorted(
-                can_delete, key=lambda e: connectors[e.connector_name].priority
-            )[-1]
-        return None
-
-    def _should_delete(self, storage_location: StorageLocation) -> bool:
-        """Get if the given StorageLocation should be deleted.
-
-        :return: True if storage_location should be deleted, False otherwise.
-        :rtype: bool
-        """
-        rules = self._get_rules(storage_location.connector_name, "delete")
-        rule_results = {}
-        rule_results["has_rules"] = len(rules) > 0
 
-        rule_results["not_locked"] = not storage_location.locked
+        default_connector_whens = [
```

Comment:
It would be better to name this `connectors_priority_whens`.

Reply:
OK.